include/drivers/eeprom.h

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2019 Vestas Wind Systems A/S
+ *
+ * Heavily based on drivers/flash.h which is:
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2016 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for EEPROM drivers
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_EEPROM_H_
+#define ZEPHYR_INCLUDE_DRIVERS_EEPROM_H_
+
+/**
+ * @brief EEPROM Interface
+ * @defgroup eeprom_interface EEPROM Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <sys/types.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef int (*eeprom_api_read)(struct device *dev, off_t offset, void *data,
+			       size_t len);
+typedef int (*eeprom_api_write)(struct device *dev, off_t offset,
+				const void *data, size_t len);
+typedef int (*eeprom_api_write_protection)(struct device *dev, bool enable);
+
+struct eeprom_driver_api {
+	eeprom_api_read read;
+	eeprom_api_write write;
+	eeprom_api_write_protection write_protection;
+	const size_t size;
+	const bool read_only;
+};
+
+/**
+ *  @brief Read data from EEPROM
+ *
+ *  @param dev EEPROM device
+ *  @param offset Address offset to read from.
+ *  @param data Buffer to store read data.
+ *  @param len Number of bytes to read.
+ *
+ *  @return 0 on success, negative errno code on failure.
+ */
+__syscall int eeprom_read(struct device *dev, off_t offset, void *data,
+			  size_t len);
+
+static inline int z_impl_eeprom_read(struct device *dev, off_t offset,
+				     void *data, size_t len)
+{
+	const struct eeprom_driver_api *api = dev->driver_api;
+
+	return api->read(dev, offset, data, len);
+}
+
+/**
+ *  @brief Write data to EEPROM
+ *
+ *  Prior to the invocation of this API, the eeprom_write_protection_set needs
+ *  to be called first to disable the write protection.
+ *
+ *  @param dev EEPROM device
+ *  @param offset Address offset to write data to.
+ *  @param data Buffer with data to write.
+ *  @param len Number of bytes to write.
+ *
+ *  @return 0 on success, negative errno code on failure.
+ */
+__syscall int eeprom_write(struct device *dev, off_t offset, const void *data,
+			   size_t len);
+
+static inline int z_impl_eeprom_write(struct device *dev, off_t offset,
+				      const void *data, size_t len)
+{
+	const struct eeprom_driver_api *api = dev->driver_api;
+
+	if (api->read_only) {
+		return -EACCES;
+	}
+
+	return api->write(dev, offset, data, len);
+}
+
+/**
+ *  @brief Enable or disable write protection for an EEPROM
+ *
+ *  @param dev EEPROM device.
+ *  @param enable enable or disable EEPROM write protection
+ *
+ *  @return 0 on success, negative errno code on failure.
+ */
+__syscall int eeprom_write_protection_set(struct device *dev, bool enable);
```

Comment:
Doesn't it make sense to make protection mechanism automatic? User know what is doing while call write API - this driver, so is no trivial like accidentally writing a register in bare-metel code.

Reply:
I think this API can be removed.

Reply:
I think this is similar to the flash idea, by separating it from the write you can have a write_protection_set, many writes, and then a write_protection_set. I think it is good to have this. If the hardware doesn't need this, it can always be a nop.

Reply:
@pabigot Your opinion?

Reply:
While I normally like explicit control of all features, the Linux at24 and at25 drivers manage WREN/WPn within the write function.

Unless there's a functional reason to require external confirmation that the write is intentional, or a significant performance impact for write-enable on each write, I agree that this function *can* be removed and protection-management done within the write function.

Whether it *should* be removed I'll leave to @henrikbrixandersen.

Reply:
@nvlsianpu, @Laczen, @pabigot: Thank you for your input on this. I have removed the write protection method from the proposed API.